feat: your dimensions accessory sidebar

HendrikThePendric opened the pull request:
Implements [TECH-877](https://jira.dhis2.org/browse/TECH-877)

---

### Key features

1. Your dimensions panel with search input
2. Extracted `DimensionsList` component which is now shared between the program and your dimension panel
3. Partially implemented list items and their interactions

---

### Description

This PR add the "Your dimensions" panel with a search input and adds some basic interactions to the dimension list items. The basic notion is to merge this PR into develop ASAP so other team member can use these dimension panels. The dimension items will be refined in a subsequent PR.

---

### Known issues

- Dimension list items don't have all their states (i.e. locked / disabled / drag-only) implemented yet (deferred to another PR). 
- Dimension list items don't show the correct icons for all dimension types yet. (not yet available at the moment).
- Selection count badge has not been implemented yet

---

### Screenshots

<img width="624" alt="Screenshot 2022-01-24 at 16 03 51" src="https://user-images.githubusercontent.com/353236/150808007-d8a6adaa-a0dc-4688-8cc4-26390725ccc6.png">

martinkrulltott commented on src/actions/ui.js:
```diff
@@ -100,9 +100,10 @@ export const tClearUi = () => (dispatch, getState) => {
 }
 
-export const acSetUiActiveModalDialog = (value) => ({
+export const acSetUiActiveModalDialog = (value, metadata) => ({
```
Unsure about this pattern of having `metadata` as an additional prop in the request here, especially since its not entirely clear what the metadata is referring to (renaming the action to `acSetUiActiveDimensionModal` might help with that). Is it preferred to use a consolidated action like this or should the consumer just do a separate "add metadata" call instead? @janhenrikoverland

HendrikThePendric replied:
> or should the consumer just do a separate "add metadata" call instead?

If this is the preferred way, we should probably remove the redux middleware.

martinkrulltott replied:
No I like the redux middleware, just want it to be clear what the metadata is referring to. So best would be to rename the action to `acSetUiActiveDimensionModal` so its obvious that its a "dimension" that is open (hence said dimension can have metadata related to it, which is passed as an optional prop). Also, whats a "ModalDialog" anyway? 😄

martinkrulltott replied:
one could also argue that "active" is too loose. "open" makes more sense (as thats what modals do, open and close, not activate and deactivate)... so `acSetUiOpenDimensionModal` 🎉

martinkrulltott commented on src/components/MainSidebar/DimensionsList/DimensionsList.js:
```diff
@@ -0,0 +1,111 @@
+        })
+    }
+    if (programName) {
+        i18n.t('No dimensions found in the {{programName}} program', {
```
```suggestion
        return i18n.t('No dimensions found in the {{programName}} program', {
```

martinkrulltott commented on src/components/MainSidebar/DimensionsList/DimensionsList.js:
```diff
@@ -0,0 +1,111 @@
+        )
+    }
+
+    if (error) {
```
technically `if` might be correct here, but when reading through a component with multiple `return`s, I think it make sense to wrap them in `if`, `if else` and finally `else` to indicate that they're indeed separate paths.

martinkrulltott replied:
same for the returns in `getNoResultsMessage` above

martinkrulltott replied:
IMO superior to all of this is to avoid multiple returns altogether, assign the value to a `let` in an `if / if else / else` block and use a single return at the end though.

HendrikThePendric replied:
This comment threw me a bit. In react components I see `if/return` it much more than `if/else`. I'm happy to comply with the code conventions of the team, but then I think the team is choosing a pattern that's a bit different from the react ecosystem as a whole....

jenniferarnesen replied:
I don't think we ever agreed on a team convention. I prefer the `if/return` as well, and if that is more compliant with the react ecosystem, then it makes sense to me to stick with it.

janhenrikoverland replied:
I understand both arguments (different paths vs react component convention).

I often think `if / else if` gets messy when the clauses are unrelated.

janhenrikoverland replied:
When checking for different values of the same varible then `if / else if` is pretty readable, like `switch / case`.

janhenrikoverland replied:
Lets stick to the react component convention in this case.

martinkrulltott replied:
I don't mind the `if/return` if it's a single use (e.g. commonly to check for a prop and return `null` to avoid multiple partial renders while loading). But if it's simply the case of checking for different conditions and returning primitive values (i18n still counts in this case) like it is with `getNoResultsMessage` I don't see why not to use `if / else if / else`.

martinkrulltott commented on src/actions/ui.js:
```diff
@@ -100,9 +100,10 @@ export const tClearUi = () => (dispatch, getState) => {
 }
 
-export const acSetUiActiveModalDialog = (value) => ({
+export const acSetUiActiveModalDialog = (value, metadata) => ({
```
```suggestion
export const acSetUiOpenDimensionModal = (value, metadata) => ({
```

martinkrulltott commented on src/components/MainSidebar/DimensionsList/DimensionsList.js:
```diff
@@ -0,0 +1,111 @@
+        )
+    }
+
+    if (dimensions.length === 0) {
```
```suggestion
    if (!dimensions?.length) {
```
`dimensions` isn't `isRequired` and has no `defaultProps` value (speaking of which, this in itself is a lint warning I think?), could thus be `undefined`?

HendrikThePendric replied:
Once loading is done, and there is no error, then we have a dimensions array. I think rendering a "no results found message" when there is a dimension array which is empty, expresses the intent of what this block of code does very well.

The suggested change makes the code more robust, but I would prefer it to fail fast.

Hope you are OK with keeping this as-is.

martinkrulltott replied:
Ok, if you want it to fail if it (for whatever reason) doesn't have a `dimension` at all then maybe that warrants a `try catch` so we can provide a user-friendly error? 
I mean, the argument "it doesn't fail now so we don't need it to be more robust" only holds up for so long, right, so the undefined-check was just a suggestion to provide some future-proofing (+ personally I think checking `!array.length` is just as expressive as `array.length === 0` so I think your intent is still very clear).

martinkrulltott commented on src/components/MainSidebar/DimensionsList/DimensionsList.js:
```diff
@@ -0,0 +1,111 @@
+    setIsListEndVisible,
+}) => {
+    const scrollBoxRef = useRef()
+    const updateIsListEndVisible = () => {
```
the name `updateIsListEndVisible` sort of clashes with `setIsListEndVisible` (but looking at it further I understand that the former is a wrapper for the latter. Was thinking it might be called `onScroll`, but that might not fit so well when called from the `useEffect` below, or does it? 🤔

HendrikThePendric replied:
I see what you mean. Tried to avoid naming collisions in 6ccaedc but not sure that's much better....

martinkrulltott replied:
haha, try reading `setIsListEndVisible(isScrollboxEndVisible(scrollBoxRef.current))` quickly 5 times in a row 😉  Hard to say, I think this way at least you can't mix up the two fns, but neither is a clear choice maybe. Up to you 👍

martinkrulltott commented on src/components/MainSidebar/DimensionsList/DimensionListItem.js:
```diff
@@ -0,0 +1,107 @@
+                <Icon />
+            </div>
+            <div className={styles.label}>{name}</div>
+            {locked ? (
```
why ternary with `''` instead of just `&&`?

HendrikThePendric replied:
Oops.... Some copy-pasted code from [the design specs](https://github.com/dhis2/design-specs/blob/master/src/components/dimension-list.js)... Well spotted. Updated in 9a65b339b9314778eb71d167bb4913ab587ff6f1

martinkrulltott commented on src/components/MainSidebar/ProgramDimensionsPanel/useProgramDimensions.js:
```diff
@@ -55,6 +61,11 @@ const reducer = (state, action) => {
                 error: action.payload,
                 dimensions: null,
             }
+        case ACTIONS.SET_LIST_END_VISIBLE:
```
Same as we discussed the other day, is `ACTIONS` ever used in a standalone fashion? Or could these props just be extract to individual consts e.g. `const ACTION_RESET = 'RESET'`?

martinkrulltott commented on src/components/MainSidebar/YourDimensionsPanel/YourDimensionsPanel.js:
```diff
@@ -0,0 +1,49 @@
+
+const YourDimensionsPanel = ({ visible }) => {
+    const [searchTerm, setSearchTerm] = useState('')
+    const debouncedSearchTerm = useDebounce(searchTerm, 500)
```
Now that `useDebounce` is used in multiple places, maybe the default debounce time should be exported as a separate prop? (e.g. `useDebounce(searchTerm, DEFAULT_DEBOUNCE_TIME)`) or let `useDebounce` set the default time internally and only pass `searchTerm` without an explicit time (e.g. `useDebounce(searchTerm)`)?

HendrikThePendric replied:
🤔 I think we might need to export two values, one for when a user is typing, and the other for when we want to respond to events that fire often (i.e. `mousemove`, `resize`, etc). Maybe like this:
```js
export const DEFAULT_USER_INPUT_DELAY = 500
export const DEFAULT_WINDOW_EVENT_DELAY = 150
```
FYI: I already used the shorter time yesterday for setting the table max-width after a window resize.

martinkrulltott replied:
Sounds good to me 👍

martinkrulltott replied:
I just saw one being `150` and figured we could have a `default` and a custom. But two "defaults" works as well

HendrikThePendric replied:
> or let `useDebounce` set the default delay internally and only pass `searchTerm` without an explicit time (e.g. `useDebounce(searchTerm)`)?

Actually this was a very nice idea. I've gone with that.

martinkrulltott commented on src/components/Visualization/useAnalyticsData.js:
```diff
@@ -136,6 +139,7 @@ const useAnalyticsData = ({
     page,
     pageSize,
 }) => {
+    const uiState = useSelector(sGetUi)
```
can we avoid passing the whole `ui` object around and only pick out the parts that are actually used? (looks like `program` and `input`)

martinkrulltott replied:
also, you already have `visualization` (which is `current`, right?), wouldn't that include the parts needed anyway?

martinkrulltott replied:
![image](https://user-images.githubusercontent.com/12590483/151166752-d5f2d8b2-284f-4f6b-9013-5fa05ff3c490.png)
debugging this I see that `program` and `input` are missing from `visualization`, but that's just because they haven't been added yet, right? so once they're added to `visualization` you could just access everything that `uiState` provides through that prop?

HendrikThePendric replied:
This was an excellent point. The fact that these values were not present in the first place was the problem, so reading them directly from the UI state was not really a solution. In 7d50f1b I have ensured that the required properties get added to `current` from `ui`.

martinkrulltott commented on src/components/Visualization/useAnalyticsData.js:
```diff
@@ -168,6 +173,7 @@ const useAnalyticsData = ({
                 mounted.current && setData({ headers, rows, pageCount, total })
                 onResponseReceived(analyticsResponse)
             } catch (error) {
+                console.error(error)
```
just left-over from debugging or is this our preferred way of logging errors?

jenniferarnesen commented:
@HendrikThePendric Should the i18n/en.pot file be in the PR?

HendrikThePendric commented:
> @HendrikThePendric Should the i18n/en.pot file be in the PR?

@jenniferarnesen There is actually 1 new string in there. The others have been moved around because I extracted a component and moved it to another folder. But yeah, we do need it in for that one string.

martinkrulltott commented on src/modules/current.js:
```diff
@@ -21,11 +21,19 @@ export const getDefaultFromUi = (current, action) => {
 }
 
+export const getProgramFromUi = (ui) =>
+    ui.program?.id ? { program: { id: ui.program.id } } : null
```
```suggestion
    ui.program?.id && { program: { id: ui.program.id } }
```

martinkrulltott replied:
same for `getProgramStageFromUi` below
